Add dash app

LukeFiddy opened the pull request.

Commit:
change test

Commit:
add plotly dependency

Commit:
add dash dependency

Commit:
add some code

Commit:
add dependency

Commit:
change dash app

Commit:
minimum working prototype

Commit:
add dependency

Commit:
add cache

Commit:
fix errors

Commit:
ruff auto fix

Commit:
change cache to json storage

Commit:
fix file storage, add download button

Commit:
fix download button visibility

Commit:
add dependency

Commit:
change table viewing

Commit:
add download pivot data button

Commit:
add dependency

Commit:
change dependency name

Commit:
add io.stringio as valid type for read_bluesky_output

Commit:
fix typing

Commit:
fix typing

Commit:
fix typing

Commit:
add test file

Commit:
change copier template to include docker

Commit:
change dockerfile

Commit:
change dockerfile

Commit:
change dockerfile

Commit:
chnage dockerfile

Commit:
change dockerfile

Commit:
change dockerfile

Commit:
change dockerfile

Commit:
change dockerfile

Commit:
change dockerfile

Commit:
change dockerfile

Commit:
change dockerfile

Commit:
change dockerfile

Commit:
change dockerfile

Commit:
change dockerfile

Commit:
change dockerfile

Commit:
change dockerfile

Commit:
change dockerfile and cli

Commit:
change dockerfile and cli

Commit:
add server options

Commit:
add server options

Commit:
add helm chart

Commit:
change formatting

Commit:
add datadict fixture

Commit:
add datadict fixture

Commit:
fix download button bug

Commit:
add tests to dash app

Commit:
fix format of strings in helm chart files

Commit:
add .helmignore to pyright ignored files

Commit:
add helm chart to exclusion on yaml check

Commit:
remove redundant code

Commit:
add indexerror test for restraints

Commit:
add server command tests

Commit:
split pages

pages are now separate, however the callback which makes the data-viewer visible when the page changes is broken

Commit:
minimum working product

Commit:
remove navbar object

Commit:
replace navigation with tabs

Commit:
change styling of page

LukeFiddy committed:
Merge branch 'main' into add_dash_app

ZohebShaikh commented on Dockerfile:
```diff
@@ -11,3 +11,26 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
+#ENTRYPOINT ["/bin/bash"]
+ENTRYPOINT [ "bimorph-mirror-analysis" ]
+CMD ["server"]
+#CMD ["ls"]
+#CMD ["/src/bimorph-mirror-analysis/app.py"]
```
nit: should be removed

LukeFiddy replied:
now removed

ZohebShaikh commented:
This looks great! The only major suggestion I have is to split [this](https://github.com/DiamondLightSource/bimorph-mirror-analysis/blob/add_dash_app/src/bimorph_mirror_analysis/app.py#L24) into two separate parts: an external_css folder and a pages folder. This will make it easier to add new pages or modify the code in the future.

Additionally, I recommend downloading the CSS files and keeping them within the project rather than relying on CDNs, as those may be updated or changed over time and aren't under our control.

For reference, here's a project that has implemented a similar approach: [Hexapod Robot Simulator](https://github.com/mithi/hexapod-robot-simulator/blob/master/app.py#L6).

ZohebShaikh commented:
I found this a bit confusing—you’re using `SemVer` for all the containers being created, but the releases follow `CalVer`. This might lead to inconsistencies or confusion down the line.

I couldn’t find any examples of containers using `CalVer` within the `diamondlightsource` org. While I know that the epics-container configs follow `CalVer`, they don’t actually contain containers.

It might be worth reconsidering this approach before merging.

Commit:
remove commented out lines from dockerfile

Commit:
add stylesheets to assets folder

Commit:
remove reference to external stylesheets

they have been downloaded and placed in the assets folder in a previous commit

LukeFiddy committed:
Merge branch 'main' into add_dash_app

Commit:
remove duplicate definition of calculate_optimal_voltages caused by rebasing main

Commit:
white space added from merge

Commit:
update_calculate_optimal_voltages and calculate_coltages for the dash app

Commit:
add test for using restrained voltage calculation method in app

Commit:
change restraints in test

Commit:
add assertion to calling calculate_voltages_with_constraints in app test

Commit:
fix check_voltages fit constraints

now covers case when the consecutive difference is too big decreasing
also added test case for this

Commit:
add side effect for mocked voltage correction calculation in app test

LukeFiddy commented:
i have downloaded the external stylesheets and added them to an assets folder. The pages issue is addressed in another branch (not yet pushed) where i have added multiple pages. I will use SemVer going forward.

ZohebShaikh commented on tests/app_test.py:
```diff
@@ -0,0 +1,192 @@
+        "Pivoted Data",
+        "Error",
+    ],
+)
+def test_change_table(data_dict: DataDict, ag_grid_selector_value: str):
+    output = change_table(data_dict, ag_grid_selector_value, {"display": "block"})
+    if ag_grid_selector_value == "Raw Data":
+        df = pd.DataFrame(data_dict["raw_data_dict"])
+        columns = [{"headerName": col, "field": col} for col in df.columns]
+        data: dict[str, str] = df.to_dict("records")  # type: ignore
+        pivot_data_style = {"display": "none"}
+        expected = (
+            columns,
+            data,
+            {"fileName": f"{data_dict['filename'].split('.')[0]}_pivoted.csv"},
+            pivot_data_style,
+        )
+
+    elif ag_grid_selector_value == "Pivoted Data":
+        df = pd.DataFrame(data_dict["pivoted_data_dict"])
+        pivot_data_style = {"display": "block", "margin": "auto"}
+        columns = [{"headerName": col, "field": col} for col in df.columns]
+        data: dict[str, str] = df.to_dict("records")  # type: ignore
+        expected = (
+            columns,
+            data,
+            {"fileName": f"{data_dict['filename'].split('.')[0]}_pivoted.csv"},
+            pivot_data_style,
+        )
+    elif ag_grid_selector_value == "Error":
+        pivot_data_style = {"display": "block", "margin": "auto"}
+        expected = ([{}], {}, {}, pivot_data_style)  # type: ignore
+
+    else:
+        raise ValueError(
+            f"value {ag_grid_selector_value} for ag_grid_selector_value parameter is \
+not covered by test"
+        )
+
+    assert output == expected
```
```suggestion
@pytest.mark.parametrize(
    "ag_grid_selector_value, data_key, pivot_data_style",
    [
        ("Raw Data", "raw_data_dict", {"display": "none"}),
        ("Pivoted Data", "pivoted_data_dict", {"display": "block", "margin": "auto"}),
        ("Error", None, {"display": "block", "margin": "auto"}),
    ],
)
def test_change_table(data_dict: DataDict, ag_grid_selector_value: str, data_key: str, pivot_data_style: dict[str, str]):
    if data_key:
        df = pd.DataFrame(data_dict.get(data_key))
        columns = [{"headerName": col, "field": col} for col in df.columns]
        data: dict[str, str] = df.to_dict("records")  # type: ignore
        expected = (
            columns,
            data,
            {"fileName": f"{data_dict['filename'].split('.')[0]}_pivoted.csv"},
            pivot_data_style,
        )
    else:  # "Error" case
        expected = ([{}], {}, {}, pivot_data_style)  # type: ignore

    output = change_table(data_dict, ag_grid_selector_value, {"display": "block"})
    assert output == expected
```

ZohebShaikh replied:
```suggestion
@pytest.mark.parametrize(
    "ag_grid_selector_value, data_key, pivot_data_style",
    [
        ("Raw Data", "raw_data_dict", {"display": "none"}),
        ("Pivoted Data", "pivoted_data_dict", {"display": "block", "margin": "auto"}),
        ("Error", None, {"display": "block", "margin": "auto"}),
    ],
)
def test_change_table(data_dict: DataDict, ag_grid_selector_value: str, data_key: str, pivot_data_style: dict[str, str]):
    if data_key:
        df = pd.DataFrame(data_dict.get(data_key))
        columns = [{"headerName": col, "field": col} for col in df.columns]
        data: dict[str, str] = df.to_dict("records")  # type: ignore
        expected = (
            columns,
            data,
            {"fileName": f"{data_dict['filename'].split('.')[0]}_pivoted.csv"},
            pivot_data_style,
        )
    else:  # "Error" case
        expected = ([{}], {}, {}, pivot_data_style)  # type: ignore

    output = change_table(data_dict, ag_grid_selector_value, {"display": "block"})
    assert output == expected
```

LukeFiddy committed:
clean up tests/app_test.py

clean up for improved readability.
moved changed variables from if statements to parameters for the test.

Co-authored-by: Zoheb Shaikh <zoheb.shaikh@diamond.ac.uk>

Commit:
clean up app_test.py:test_change_table

ZohebShaikh commented on tests/app_test.py:
```diff
@@ -0,0 +1,177 @@
+    assert output == expected
+
+
+@pytest.mark.parametrize("n_clicks", [1, None])
+def test_export_data_as_csv(n_clicks: int | None):
+    output = export_data_as_csv(n_clicks)
+    if n_clicks:  # only when int
+        assert output  # should be true
+    else:  # when Nnoe
+        assert not output  # should be false
```
```suggestion
@pytest.mark.parametrize("n_clicks, expected", [(1, True), (None, False)])
def test_export_data_as_csv(n_clicks: int | None, expected: bool):
    assert export_data_as_csv(n_clicks) == expected
```

ZohebShaikh commented:
Only few changes :-
1. nit: can change tests for app_test to test_app...etc It is a unwritten convention in all other repos (Can have a look at [blueapi](https://github.com/DiamondLightSource/blueapi/tree/main/tests/unit_tests/service),dodal,ophyd_async,etc)
2. nit: Can you add where the css where downloaded form in the top of those files (the cdn url)

ZohebShaikh commented:
I've reviewed it. @dan-fernandes, could you take a look as well?

LukeFiddy committed:
Update tests/app_test.py

clean up test_export_data_as_csv

Co-authored-by: Zoheb Shaikh <zoheb.shaikh@diamond.ac.uk>

Commit:
add comments to css stylesheets containing info on their download source

LukeFiddy committed:
Merge branch 'main' into add_dash_app

LukeFiddy committed:
Merge branch 'add_dash_app' into add_multiple_pages_to_dash

LukeFiddy committed:
Merge pull request #42 from DiamondLightSource/add_multiple_pages_to_dash

Add multiple pages to dash
